src/components/select/SelectWithPredicate.tsx

```diff
@@ -2,43 +2,76 @@ import * as React from 'react';
-import {ReduxConnect} from '../../utils/ReduxUtils';
+import {callIfDefined} from '../../utils/FalsyValuesUtils';
+import {IDispatch, ReduxConnect} from '../../utils/ReduxUtils';
 import {FlatSelectConnected} from '../flatSelect/FlatSelectConnected';
 import {IFlatSelectOptionProps} from '../flatSelect/FlatSelectOption';
 import {IItemBoxProps} from '../itemBox/ItemBox';
-import {ISelectProps} from './SelectConnected';
+import {updateListBoxOption} from '../listBox/ListBoxActions';
+import {ISelectWithPredicateAndFilterProps} from './SelectComponents';
+import {ISelectOwnProps, ISelectProps, ISelectStateProps} from './SelectConnected';
+
+export type MatchPredicate = (predicate: string, item: IItemBoxProps) => boolean;
 
 export interface ISelectWithPredicateOwnProps {
     options: IFlatSelectOptionProps[];
-    matchPredicate: (predicate: string, item: IItemBoxProps) => boolean;
+    matchPredicate?: MatchPredicate;
 }
-const SelectWithPredicatePropsToOmit = keys<ISelectWithPredicateOwnProps>();
-
-export interface ISelectWithPredicateProps extends ISelectWithPredicateOwnProps, ISelectProps {}
 
-export const selectWithPredicate = (Component: (React.ComponentClass<ISelectProps> | React.StatelessComponent<ISelectProps>)): React.ComponentClass<ISelectWithPredicateProps> => {
+export interface ISelectWithPredicateDispatchProps {
+    updateItems: (predicate: string, items: IItemBoxProps[], matchPredicate: MatchPredicate) => void;
+}
 
-    const mapStateToProps = (state: IReactVaporState, ownProps: ISelectWithPredicateProps): Partial<ISelectProps> => {
-        const flatSelect = _.findWhere(state.flatSelect, {id: ownProps.id});
-        const predicate = flatSelect && flatSelect.selectedOptionId || ownProps.options[0].id;
+export interface ISelectWithPredicateProps extends ISelectWithPredicateOwnProps,
+    Partial<ISelectWithPredicateDispatchProps>,
+    ISelectProps {}
 
-        const items = _.map(ownProps.items, (item: IItemBoxProps) => {
-            const visible = ownProps.matchPredicate(predicate, item);
+const SelectWithPredicatePropsToOmit = keys<ISelectWithPredicateOwnProps>();
 
-            return {...item, hidden: !visible || item.hidden};
-        });
+export const selectWithPredicate = (Component: (React.ComponentClass<ISelectProps> | React.StatelessComponent<ISelectProps>)): React.ComponentClass<ISelectWithPredicateAndFilterProps> => {
 
+    const mapStateToProps = (state: IReactVaporState, ownProps: ISelectWithPredicateProps): Partial<ISelectStateProps> => {
+        const listbox = _.findWhere(state.listBoxes, {id: ownProps.id});
         return {
-            items,
+            items: listbox ? listbox.items : ownProps.items,
         };
     };
 
-    @ReduxConnect(mapStateToProps)
-    class WrappedComponent extends React.Component<ISelectWithPredicateProps> {
+    const mapDispatchToProps = (dispatch: IDispatch, ownProps: ISelectWithPredicateOwnProps & ISelectOwnProps): ISelectWithPredicateDispatchProps => ({

```

Comment:
i think this could be simplified and still be very readable:  

```
const mapDispatchToProps = (dispatch: IDispatch, ownProps: ISelectWithPredicateOwnProps & ISelectOwnProps): ISelectWithPredicateDispatchProps => ({
    updateItems: (predicate: string, items: IItemBoxProps[], matchPredicate: MatchPredicate) => dispatch(updateListBoxOption(
        ownProps.id,
        _.map(items, (item: IItemBoxProps) => ({...item, hidden: !matchPredicate(predicate, item)})),
    )),
});
```